alt_e2eshark/e2e_testing/backends.py

```diff
@@ -0,0 +1,41 @@
+# This file will contain customizations for how to compile mlir from various entrypoints
+import abc
+from typing import TypeVar
+
+CompiledArtifact = TypeVar("CompiledArtifact")
+Invoker = TypeVar("Invoker")
+
+class BackendBase(abc.ABC):
+
+    @abc.abstractmethod
+    def compile(self, module) -> CompiledArtifact:
+        '''specifies how to compile an MLIR Module'''
+
+    @abc.abstractmethod
+    def load(self, artifact: CompiledArtifact) -> Invoker:
+        '''loads the compiled artifact'''
+
+
+from iree import compiler as ic
+from iree import runtime as rt
+from torch_mlir.passmanager import PassManager
+
+class SimpleIREEBackend(BackendBase):
+
+    def compile(self, module):
+        pipeline = "builtin.module(func.func(convert-torch-onnx-to-torch))"
+        with module.context as ctx:
+            pm = PassManager.parse(pipeline)
+            pm.run(module.operation)
+        return ic.tools.compile_str(str(module), input_type="torch", target_backends=["llvm-cpu"])
```

Comment:
> > avoid running command-line scripts through python. This includes transitioning to using python bindings for torch-mlir and iree, and developing an import-friendly alternative to model.py files.
> 
> A few thoughts here:
> 
>     * For iree_tests, running command-line scripts is an explicit design goal, so that developers working on the native tools can run (and debug) the tests directly (step by step) and the tests can be run on platforms without a python interpreter. The test runner there uses pytest and is thus written in Python, but other test runners could be written using CMake/CTest, Bazel, etc.

I see. It didn't make sense to me to run command line scripts from python if equivalent python bindings exist, since the bindings probably have a better api than organizing a command-line string. I suppose it is still possible with the class structure here to define a backend which uses command line scripting, if that is a design choice that we want to keep here. Either way I'd like to still dump all of the relevant files in an organized location, and have a log printing out equivalent command-line reproducers.

I guess if we wanted something python-agnostic, it might be best to upload the onnx files, and input.bin files, mlir to some storage that people could access if they wanted to run tests without python? 

> 
>     * For e2eshark, using Python bindings like here instead of native tools is a reasonable choice, though I'd keep a close eye on how to generate useful errors / logs / dumps. One of the first things that a compiler developer typically wants from a failing test case is a standalone reproducer.
> 
>     * In general, test suites should aim to use projects as users of those projects would. If the Python bindings prove difficult to use for whatever reason, that's something this sort of work could help highlight and drive improvements on.

Yeah, I'm finding the iree python bindings a bit difficult to wrap my head around right now, so when I get to know them a bit better I'll see if I have any suggestions to make.

Reply:
> I guess if we wanted something python-agnostic, it might be best to upload the onnx files, and input.bin files, mlir to some storage that people could access if they wanted to run tests without python?

That's what iree_tests does (sort of, it starts from .mlir and has .onnx as a preprocessing step run offline). I've been thinking that model developers and people working on torch-mlir can use e2eshark as a hackable testing space then tests are exported from e2eshark into iree_tests. I have the start of a script to do such an export/import at https://github.com/nod-ai/SHARK-TestSuite/blob/main/iree_tests/pytorch/models/import_from_e2eshark.py. Having that sort of separation is also why I'm fine with making a different design choice here.